### PR TITLE
Upgrade samples

### LINKED ISSUE #2089: QA Day: Samples
* Check that all samples run with the current Espresso
* Put doc strings at the top of the sample files explaining what they do
* Put a list of required Espresso features in the samples and check for them

required_features = []
espressomd.assert_features(required_features)

### PR DESCRIPTION
**Pull request by jngrad**
Fixes #2089

Description of changes:
- remove unused variables/imports/functions and some of the commented out code
- remove Tcl-specific code, replace manual warmup by steepest descent
- use `argparse` to process user input and document the parameters, stop printing features
- improve docstrings and use them to generate an always up-to-date sample list in the docs

### CODE REVIEW

File: samples/lj-demo.py [hunk 1]
```diff
@@ -132,7 +136,7 @@
 system.analysis.dist_to(0)
 
 act_min_dist = system.analysis.min_dist()
-system.cell_system.max_num_cells = 2744
+system.cell_system.max_num_cells = 14**3
```
Comment (KaiSzuttor) [inline samples/lj-demo.py]: where does this magic number come from?

File: samples/lj_liquid.py [hunk 2]
```diff
@@ -45,7 +48,7 @@
 box_l = 10.7437
```
Comment (KaiSzuttor) [inline samples/lj_liquid.py:47]: please remove the outdated comment above this line

File: samples/lj_liquid.py [hunk 3]
```diff
@@ -93,21 +96,21 @@
...
 
-print("Simulate {} particles in a cubic simulation box of length {} at density {}."
+print("Simulate {} particles in a cubic box of length {} at density {}."
       .format(n_part, box_l, density).strip())
 print("Interactions:\n")
 act_min_dist = system.analysis.min_dist()
 print("Start with minimal distance {}".format(act_min_dist))
 
-system.cell_system.max_num_cells = 2744
+system.cell_system.max_num_cells = 14**3
```
Comment (KaiSzuttor) [inline samples/lj_liquid.py]: magic number?

File: samples/lj_liquid.py [hunk 4]
```diff
@@ -133,22 +136,16 @@
 
 # Warmup Integration Loop
 i = 0
-while (i < warm_n_times and act_min_dist < min_dist):
+while i < warm_n_times and act_min_dist < min_dist:
```
Comment (KaiSzuttor) [inline samples/lj_liquid.py]: maybe we should ```minimize_energy``` here

File: samples/lj_liquid_distribution.py [hunk 5]
```diff
@@ -48,7 +45,7 @@
 box_l = 10.7437
```
Comment (KaiSzuttor) [inline samples/lj_liquid_distribution.py:44]: please remove the outdated comment above this line

File: samples/lj_liquid_distribution.py [hunk 6]
```diff
@@ -125,13 +120,13 @@
...
 
-print("Simulate {} particles in a cubic simulation box of length {} at density {}."
+print("Simulate {} particles in a cubic box of length {} at density {}."
       .format(n_part, box_l, density).strip())
 print("Interactions:\n")
 act_min_dist = system.analysis.min_dist()
 print("Start with minimal distance {}".format(act_min_dist))
 
-system.cell_system.max_num_cells = 2744
+system.cell_system.max_num_cells = 14**3
```
Comment (KaiSzuttor) [inline samples/lj_liquid_distribution.py]: magic number?

File: samples/lj_liquid_structurefactor.py [hunk 7]
```diff
@@ -114,13 +118,13 @@
...
 
-print("Simulate {} particles in a cubic simulation box {} at density {}."
+print("Simulate {} particles in a cubic box {} at density {}."
       .format(n_part, box_l, density).strip())
 print("Interactions:\n")
 act_min_dist = system.analysis.min_dist()
 print("Start with minimal distance {}".format(act_min_dist))
 
-system.cell_system.max_num_cells = 2744
+system.cell_system.max_num_cells = 14**3
```
Comment (KaiSzuttor) [inline samples/lj_liquid_structurefactor.py]: magic number?

File: samples/lj_liquid_structurefactor.py [hunk 8]
```diff
@@ -143,18 +147,16 @@
 
 # Warmup Integration Loop
 i = 0
-while (i < warm_n_times and act_min_dist < min_dist):
+while i < warm_n_times and act_min_dist < min_dist:
```
Comment (KaiSzuttor) [inline samples/lj_liquid_structurefactor.py]: ```minimize_energy```?

File: samples/minimal-charged-particles.py [hunk 9]
```diff
@@ -92,7 +92,7 @@
 system.force_cap = wca_cap
 i = 0
 act_min_dist = system.analysis.min_dist()
-while (i < warm_n_times and act_min_dist < min_dist):
+while i < warm_n_times and act_min_dist < min_dist:
```
Comment (KaiSzuttor) [inline samples/minimal-charged-particles.py]: ```minimize_energy```?

File: samples/minimal-diamond.py [hunk 10]
```diff
@@ -90,15 +84,14 @@
 system.thermostat.set_langevin(kT=0.0, gamma=1.0)
 
 # slowly ramp up the cap
-while (wca_cap < 5):
+while wca_cap < 5:
```
Comment (KaiSzuttor) [inline samples/minimal-diamond.py]: ```minimize_energy```?

File: samples/minimal-polymer.py [hunk 11]
```diff
@@ -67,26 +65,24 @@
...
 system.force_cap = wca_cap
-i = 0
 act_min_dist = system.analysis.min_dist()
 
 # warmup with zero temperature to remove overlaps
 system.thermostat.set_langevin(kT=0.0, gamma=1.0)
 
 # slowly ramp un up the cap
-while (act_min_dist < 0.95):
+while act_min_dist < 0.95:
```
Comment (KaiSzuttor) [inline samples/minimal-polymer.py]: ```minimize_energy```?

File: samples/p3m.py [hunk 12]
```diff
@@ -159,7 +175,7 @@
 
 # Warmup Integration Loop
 i = 0
-while (i < warm_n_times or act_min_dist < min_dist):
+while i < warm_n_times or act_min_dist < min_dist:
```
Comment (KaiSzuttor) [inline samples/p3m.py]: ```minimize_energy```?

File: samples/reaction_ensemble.py [hunk 13]
```diff
@@ -53,7 +54,7 @@
 
 system.time_step = 0.02
 system.cell_system.skin = 0.4
-system.cell_system.max_num_cells = 2744
+system.cell_system.max_num_cells = 14**3
```
Comment (KaiSzuttor) [inline samples/reaction_ensemble.py]: magic number?

File: samples/rigid_body.py [hunk 14]
```diff
@@ -16,21 +16,19 @@
...
-from espressomd import integrate
 from espressomd.virtual_sites import VirtualSitesRelative
 
 import numpy as np
 
 
 box_l = 100
-system = espressomd.System(box_l=[box_l, box_l, box_l])
+system = espressomd.System(box_l=3 * [box_l])
 system.set_random_state_PRNG()
```
Comment (KaiSzuttor) [inline samples/rigid_body.py]: this is almost never needed (unless you use NPT)

File: samples/slice_input.py [hunk 15]
```diff
@@ -47,7 +44,7 @@
 system.time_step = 0.01
 system.cell_system.skin = 0.4
 
-system.cell_system.max_num_cells = 2744
+system.cell_system.max_num_cells = 14**3
```
Comment (KaiSzuttor) [inline samples/slice_input.py]: magic number?

File: samples/visualization_ljliquid.py [hunk 16]
```diff
@@ -136,36 +137,16 @@
 
 # Warmup Integration Loop
 i = 0
-while (i < warm_n_times and act_min_dist < min_dist):
+while i < warm_n_times and act_min_dist < min_dist:
```
Comment (KaiSzuttor) [inline samples/visualization_ljliquid.py]: ```minimize_energy```?

File: samples/visualization_ljliquid.py [hunk 17]
```diff
@@ -186,12 +167,10 @@
 pyplot.legend()
 pyplot.show(block=False)
 
-j = 0
-
 
 def main_loop():
     global energies
-    print("run %d at time=%f " % (i, system.time))
+    print("run at time=%f " % system.time)
```
Comment (KaiSzuttor) [inline samples/visualization_ljliquid.py]: we should probably also use ```"{}".format(```

File: samples/visualization_poiseuille.py [hunk 18]
```diff
@@ -15,20 +15,20 @@
...
 import espressomd.visualization_opengl
 
 required_features = ["LB_BOUNDARIES", "EXTERNAL_FORCES"]
 espressomd.assert_features(required_features)
 
 # System setup
 box_l = 16
-system = System(box_l=[box_l] * 3)
+system = espressomd.System(box_l=[box_l] * 3)
 system.set_random_state_PRNG()
```
Comment (KaiSzuttor) [inline samples/visualization_poiseuille.py]: not needed

File: samples/widom_insertion.py [hunk 19]
```diff
@@ -48,7 +43,7 @@
 
 # Integration parameters
 #############################################################
-system = espressomd.System(box_l=[box_l, box_l, box_l])
+system = espressomd.System(box_l=3 * [box_l])
 system.set_random_state_PRNG()
```
Comment (KaiSzuttor) [inline samples/widom_insertion.py]: not needed

File: samples/drude_bmimpf6.py [hunk 20]
```diff
@@ -75,7 +72,7 @@
 box_l = box_volume**(1. / 3.)
 print("\n-->Ion pairs:", n_ionpairs, "Box size:", box_l)
 
-system = espressomd.System(box_l=[box_l, box_l, box_l])
+system = espressomd.System(box_l=3 * [box_l])
 system.virtual_sites = VirtualSitesRelative(have_velocity=True)
 system.set_random_state_PRNG()
```
Comment (KaiSzuttor) [inline samples/drude_bmimpf6.py]: not needed

File: samples/ekboundaries.py [hunk 21]
```diff
@@ -15,19 +15,18 @@
...
 espressomd.assert_features(required_features)
 
-from espressomd import System, shapes, electrokinetics, ekboundaries
-import sys
+from espressomd import shapes, electrokinetics, ekboundaries
 import os
 
-system = System(box_l=[10, 10, 10])
+system = espressomd.System(box_l=[10, 10, 10])
 system.set_random_state_PRNG()
```
Comment (KaiSzuttor) [inline samples/ekboundaries.py]: not needed

File: samples/grand_canonical.py [hunk 22]
```diff
@@ -55,7 +58,7 @@
 
 # Integration parameters
 #############################################################
-system = espressomd.System(box_l=[box_l, box_l, box_l])
+system = espressomd.System(box_l=3 * [box_l])
 system.set_random_state_PRNG()
```
Comment (KaiSzuttor) [inline samples/grand_canonical.py]: probably not needed

File: samples/lj-demo.py [hunk 23]
```diff
@@ -88,7 +94,7 @@
 
 # Integration parameters
 #############################################################
-system = espressomd.System(box_l=[1.0, 1.0, 1.0])
+system = espressomd.System(3 * [box_l])
 system.set_random_state_PRNG()
```
Comment (KaiSzuttor) [inline samples/lj-demo.py:98]: not needed

**Review by KaiSzuttor — DISMISSED** [review]

File: samples/lj_liquid.py (hunk 4 above)
Comment (RudolfWeeber) [reply samples/lj_liquid.py]: If we should, we should system.integrator.set_steepest_descent()

File: samples/lj-demo.py (hunk 1 above)
Comment (jngrad) [reply samples/lj-demo.py]: I've removed all occurrences of `max_num_cells`. There was no performance impact.

File: samples/lj_liquid.py (hunk 4 above)
Comment (jngrad) [reply samples/lj_liquid.py]: Done in 23d39c7 for all samples.

File: doc/sphinx/io.rst [hunk 24]
```diff
@@ -158,35 +159,39 @@ Currently available:
...
 .. code:: python
     
     import h5py
     h5file = h5py.File("h5mdfile.h5", 'r')
     positions = h5file['particles/atoms/position/value']
 
-Further the files can be inspected with the GUI tool hdfview.
+Furthermore, the files can be inspected with the GUI tool hdfview.
+
+For other examples, see :file:`/samples/5hmd.py`
```
Comment (KaiSzuttor) [inline doc/sphinx/io.rst]: typo?

File: samples/electrophoresis.py [hunk 25]
```diff
@@ -17,30 +17,26 @@
...
 
-from espressomd import thermostat
 from espressomd import interactions
 from espressomd import electrostatics
-import sys
 import numpy as np
 try:
     import cPickle as pickle
 except ImportError:
     import pickle
```
Comment (KaiSzuttor) [inline samples/electrophoresis.py]: I think we should remove pickle here and also the lattice Boltzmann stuff. Either we show a sample with hydrodynamics and explain it or we don't. Checkpointing is also not part of this sample.

File: doc/sphinx/io.rst (hunk 24 above)
Comment (jngrad) [reply doc/sphinx/io.rst]: yes

File: samples/electrophoresis.py (hunk 25 above)
Comment (jngrad) [reply samples/electrophoresis.py]: agreed, let's remove pickle and lb

File: samples/drude_bmimpf6.py [hunk 26]
```diff
@@ -16,16 +16,13 @@
 # along with this program.  If not, see <http://www.gnu.org/licenses/>.
 """
 Particle polarization with cold Drude oscillators on a coarse grained
```
Comment (pkreissl) [inline samples/drude_bmimpf6.py]: Being pedantic, "coarse-grained simulation", I guess..

File: samples/electrophoresis.py [hunk 27]
```diff
@@ -139,41 +124,38 @@
...
+
+# activate thermostat
+system.thermostat.set_langevin(kT=1.0, gamma=1.0, seed=42)
 
 #############################################################
 #      Sampling                                             #
 #############################################################
 #
-# Activate electrostatic with checkpoint example
+# Activate electrostatic
```
Comment (pkreissl) [inline samples/electrophoresis.py]: electrostatic**s**

File: samples/grand_canonical.py [hunk 28]
```diff
@@ -17,16 +17,19 @@
 # along with this program.  If not, see <http://www.gnu.org/licenses/>.
 #
 """
-This example script performs a grand canonical simulation of a system in contact
+Perform a grand canonical simulation of a system in contact
 with a salt reservoir and ensures constant chemical potential.
```
Comment (pkreissl) [inline samples/grand_canonical.py]: [...] and ensure~s~ [...] / while maintaining a constant / or similar

File: samples/lj_liquid.py [hunk 29]
```diff
@@ -17,59 +17,58 @@
...
-This sample simulates a Lennard-Jones fluid maintained at a fixed temperature
-by a Langevin thermostat.
+Simulate a Lennard-Jones fluid maintained at a fixed temperature
+by a Langevin thermostat. Shows the basic features of how to:
+
+* set up system parameters, particles and interactions.
+* warm up and integrate.
+* write parameters, configurations and observables to files
+
+The particles in the system are of two types: Type 0 and Type 1.
```
Comment (pkreissl) [inline samples/lj_liquid.py]: **t**ype 0 and **t**ype 1

File: samples/lj_liquid.py [hunk 30]
```diff
@@ -17,59 +17,58 @@
...
+Type 0 particles interact with each other via a repulsive WCA
+interaction. Type 1 particles neither interact with themselves
+nor with Type 0 particles.
```
Comment (pkreissl) [inline samples/lj_liquid.py]: [...] nor with **t**ype 0 [...]

File: samples/lj_liquid_distribution.py [hunk 31]
```diff
@@ -17,44 +17,39 @@
...
 #
 """
-This sample sets up a Lennard-Jones fluid. The particles in the system are of
-two types: Type 0 and Type 1. Type 0 particles interact with each other via a
-repulsive WCA interaction. Type 1 particles neither interact with themselves
-nor with Type 0 particles. The distribution of minimum distances between
-particles of Type 0 and Type 1 is recorded.
+Set up a Lennard-Jones fluid maintained at a fixed temperature by a
+Langevin thermostat. The particles in the system are of two types:
+Type 0 and Type 1. Type 0 particles interact with each other via a
```
Comment (pkreissl) [inline samples/lj_liquid_distribution.py]: [...] **t**ype 0 and **t**ype 1.

File: samples/lj_liquid_distribution.py [hunk 32]
```diff
@@ -17,44 +17,39 @@
...
+repulsive WCA interaction. Type 1 particles neither interact with
+themselves nor with Type 0 particles. The distribution of minimum
```
Comment (pkreissl) [inline samples/lj_liquid_distribution.py]: **t**ype 0

File: samples/lj_liquid_distribution.py [hunk 33]
```diff
@@ -17,44 +17,39 @@
...
+distances between particles of Type 0 and Type 1 is recorded with
```
Comment (pkreissl) [inline samples/lj_liquid_distribution.py]: **t**ype 0 and **t**ype 1

File: samples/lj_liquid_structurefactor.py [hunk 34]
```diff
@@ -16,37 +16,40 @@
 # You should have received a copy of the GNU General Public License
 # along with this program.  If not, see <http://www.gnu.org/licenses/>.
 #
+"""
+Set up a Lennard-Jones fluid maintained at a fixed temperature by a
+Langevin thermostat. The particles in the system are of two types:
+Type 0 and Type 1. Type 0 particles interact with each other via a
```
Comment (pkreissl) [inline samples/lj_liquid_structurefactor.py]: same as above

File: samples/lj_liquid_structurefactor.py [hunk 35]
```diff
@@ -16,37 +16,40 @@
...
+repulsive WCA interaction. Type 1 particles neither interact with
+themselves nor with Type 0 particles. The spherically averaged
+structure factor of particles of Type 0 and Type 1 is calculated
```
Comment (pkreissl) [inline samples/lj_liquid_structurefactor.py]: again

File: samples/load_checkpoint.py [hunk 36]
```diff
@@ -15,7 +15,13 @@
...
 # along with this program.  If not, see <http://www.gnu.org/licenses/>.
 """
-This sample illustrates how various observables of interest can be checkpointed.
+Basic usage of the checkpointing feature. Show how to load the state of:
+
+* custom user variables
+* non-bonded interactions
+* particles
+* P3M parameters
+* thermostat
```
Comment (pkreissl) [inline samples/load_checkpoint.py]: Missing the full stops at the end of each item.

File: samples/object_in_fluid/motivation.py [hunk 37]
```diff
@@ -15,35 +15,39 @@
 # You should have received a copy of the GNU General Public License
 # along with this program.  If not, see <http://www.gnu.org/licenses/>.
 """
-This sample loads model red blood cells and simulates its motion
-in a complex geometry.
+Simulate the motion of flexible red blood cells in a lattice-Boltzmann fluid
```
Comment (pkreissl) [inline samples/object_in_fluid/motivation.py:18]: Actually this is exactly how I would write "lattice-Boltzmann", i.e. with a lower-case "l". However everywhere else in the samples, an upper-case "L" is used. I would very much prefer consistency...

File: samples/reaction_ensemble.py [hunk 38]
```diff
@@ -17,21 +17,22 @@
...
-in which ensemble you want to simulate via either providing --reaction_ensemble or --constant_pH_ensemble as command line argument to the script.
-Be aware that in the case of the reaction ensemble the dissociation constant gamma is not the thermodynamic reaction constant K, but rather K*1mol/l and therefore carries a unit!.
-In the case of the of the constant pH method gamma is the thermodynamic reaction constant!
+Guide for the reaction ensemble and the constant pH ensemble.
+"""
+epilog = """You can choose in which ensemble you want to simulate via either
+providing --reaction_ensemble or --constant_pH_ensemble as command line
+argument to the script. Be aware that in the case of the reaction ensemble,
+the dissociation constant gamma is not the thermodynamic reaction constant K,
+but rather K*1mol/l and therefore carries a unit! In the case of the of the
```
Comment (pkreissl) [inline samples/reaction_ensemble.py]: Can be left as is, would prefer "K * 1 mol/l". Your call.

File: samples/save_checkpoint.py [hunk 39]
```diff
@@ -15,7 +15,13 @@
...
 # along with this program.  If not, see <http://www.gnu.org/licenses/>.
 """
-This sample demonstrates how to checkpoint a simulation.
+Basic usage of the checkpointing feature. Show how to write the state of:
+
+* custom user variables
+* non-bonded interactions
+* particles
+* P3M parameters
+* thermostat
```
Comment (pkreissl) [inline samples/save_checkpoint.py]: Missing full stops.

File: samples/visualization_interactive.py [hunk 40]
```diff
@@ -15,15 +15,13 @@
 # You should have received a copy of the GNU General Public License
 # along with this program.  If not, see <http://www.gnu.org/licenses/>.
 """
-Visualization samples with keyboard control and particle repositioning via
-timed callback.
+Visualize a simulation box where the particles can be repositioned via the
+mouse and timed callbacks, and the temperature of the thermostat changed
+via the keyboard.
```
Comment (pkreissl) [inline samples/visualization_interactive.py]: Maybe another [...] of the thermostat "can be" changed [...] for immediate clarity. Your call.

File: samples/visualization_lbboundaries.py [hunk 41]
```diff
@@ -15,7 +15,7 @@
 # You should have received a copy of the GNU General Public License
 # along with this program.  If not, see <http://www.gnu.org/licenses/>.
 """
-Visualization sample for Lattice Boltzmann boundary nodes.
+Visualize lattice-Boltzmann boundary nodes.
```
Comment (pkreissl) [inline samples/visualization_lbboundaries.py:18]: See comment on upper-/lower-case "l" above.

File: samples/visualization_poiseuille.py [hunk 42]
```diff
@@ -15,12 +15,12 @@
 # You should have received a copy of the GNU General Public License
 # along with this program.  If not, see <http://www.gnu.org/licenses/>.
 """
-Visualization sample for Poiseuille flow with Lattice Boltzmann.
+Visualize the Poiseuille flow in a lattice-Boltzmann fluid with an
```
Comment (pkreissl) [inline samples/visualization_poiseuille.py:18]: Again, I fully approve of "lattice", but please be consistent...

**Review by pkreissl — CHANGES_REQUESTED** [review]
Content-wise nice clean-up. Lgtm, now that the points raised by @KaiSzuttor have been addressed.
I only found a few more or (being a bit pedantic) less important ortho/typographical issues.

**Review by pkreissl — APPROVED** [review]

**Review by RudolfWeeber — APPROVED** [review]

**Pull request merged**